### PR TITLE
Clarify the PEM docs

### LINKED ISSUE #16966: Can't pass a library context to PEM_read_X509()
I'm making the following sequence of API calls to OpenSSL:

    ctx = SSL_CTX_new_ex(non_fips_libctx, NULL, TLS_method());
    SSL_CTX_use_PrivateKey_file(ctx,<keyfile>,SSL_FILETYPE_PEM);
    SSL_CTX_use_certificate_file(ctx,<certfile>,SSL_FILETYPE_PEM);
    SSL_CTX_check_private_key(ctx);
    fp = fopen(<certfile>, "r");
    mycert = PEM_read_X509(fp, NULL, 0, NULL);
    pkey = X509_get_pubkey(mycert);

OpenSSL throws the following error on the call to X509_get_pubkey():

    00079FF8647F0000:error:03000072:digital envelope routines:(unknown function):decode error:crypto/x509/x_pubkey.c:444:

This is possibly because the call to PEM_read_X509() does not take a library context as an argument. My applications sets up two non-default library contexts, one for FIPS (loading the fips and base providers), and one for non-FIPS (loads the default provider). No providers are loaded into the default library context.

Details from Matt Caswell:

"The public key is actually decoded as part of the PEM_read_X509(). But,
the PEM_read_X509() call does not take a libctx parameter. Subsequently,
it attempts to decode the embedded X509 public key and will use the
default libctx - which has no providers and so the decode of the key
fails. Therefore when you subsequently try to obtain the public key with
X509_get_pubkey() you get a NULL return."

See more information about this problem here:

https://www.mail-archive.com/openssl-users@openssl.org/msg90228.html

and here:

https://www.mail-archive.com/openssl-users@openssl.org/msg90242.html

I'm using the following OpenSSL version:

    OpenSSL 3.0.0 7 sep 2021 (Library: OpenSSL 3.0.0 7 sep 2021)

### PR DESCRIPTION
**Pull request by mattcaswell**
Make it clear how the existing PEM functions can be used to create an
X509, X509_REQ or X509_CRL object with an associated libctx/propq.

Fixes #16966

### CODE REVIEW

File: doc/man3/PEM_read_bio_PrivateKey.pod [hunk 1]
```diff
@@ -448,7 +455,8 @@ where I<x> already contains a valid certificate, may not work, whereas:
  X509_free(x);
  x = PEM_read_bio_X509(bp, NULL, 0, NULL);
 
-is guaranteed to work.
+is guaranteed to work. It is always acceptable for I<x> to contain a newly
+allocacted, empty B<X509> object (for example allocated via L<X509_new_ex(3)>).
```
Comment (t8m) [inline doc/man3/PEM_read_bio_PrivateKey.pod:459]: Typo: allocated

**Comment by Euneverinc** [comment]
Please unsubscribe me to this e-mail feed and e-Mail address immediately!
Thank you.

Liz cahall

**Comment by mattcaswell** [comment]
> Please unsubscribe me to this e-mail feed and e-Mail address immediately!

We can't do that for you. You need to unsubscribe yourself. See:
https://docs.github.com/en/account-and-profile/managing-subscriptions-and-notifications-on-github/managing-subscriptions-for-activity-on-github/managing-your-subscriptions

File: doc/man3/PEM_read_bio_PrivateKey.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/man3/PEM_read_bio_PrivateKey.pod]: Fixed

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed to master and 3.0. Thanks.

**Pull request closed without merging**